polkadot/node/core/prospective-parachains/src/lib.rs

```diff
@@ -430,81 +455,121 @@ async fn handle_candidate_introduced<Context>(
 				"Received seconded candidate for inactive para",
 			);
 
-			let _ = tx.send(Vec::new());
-			return Ok(())
+			let _ = tx.send(false);
+			return
 		},
 		Some(storage) => storage,
 	};
 
-	let candidate_hash = match storage.add_candidate(candidate, pvd) {
-		Ok(c) => c,
-		Err(CandidateStorageInsertionError::CandidateAlreadyKnown(c)) => {
-			// Candidate known - return existing fragment tree membership.
-			let _ = tx.send(fragment_tree_membership(&view.active_leaves, para, c));
-			return Ok(())
-		},
-		Err(CandidateStorageInsertionError::PersistedValidationDataMismatch) => {
-			// We can't log the candidate hash without either doing more ~expensive
-			// hashing but this branch indicates something is seriously wrong elsewhere
-			// so it's doubtful that it would affect debugging.
+	let actual_relay_parent = candidate.descriptor.relay_parent;
+	let parent_head_hash = pvd.parent_head.hash();
+	let output_head_hash = Some(candidate.commitments.head_data.hash());
+	let candidate_hash =
+		match storage.add_candidate(candidate.clone(), pvd, CandidateState::Seconded) {
```

Comment:
It seems a bit counter intuitive to why we need to add the candidate to storage then remove it. It would be better if we do some proper renaming and adding some comments about how we use the candidate storage.

Reply:
may be a bit counter-intuitive at first, but I did this to avoid duplicating a lot of code between the `populate` and `repopulate` methods.

I'll rename the repopulate method as you suggested and add more comments here

Reply:
done